### PR TITLE
fix `devices.qubit.apply_operation` with large tf states

### PR DESCRIPTION
**Pull request by albi3ro**
Tensorflow doesn't support slicing on states with more than 8 dimensions. So if we try and run the code:

```
state = create_initial_state(range(10), like='tensorflow')
apply_operation(qml.PauliZ(9), state)
```

We get:
```
---------------------------------------------------------------------------
UnimplementedError                        Traceback (most recent call last)
Cell In[7], line 2
      1 state = create_initial_state(range(10), like='tensorflow')
----> 2 apply_operation(qml.PauliZ(9), state)

File Python.framework/Versions/3.10/lib/python3.10/functools.py:889, in singledispatch.<locals>.wrapper(*args, **kw)
    885 if not args:
    886     raise TypeError(f'{funcname} requires at least '
    887                     '1 positional argument')
--> 889 return dispatch(args[0].__class__)(*args, **kw)

File pennylane/pennylane/devices/qubit/apply_operation.py:183, in apply_pauliz(op, state)
    180 sl_0 = _get_slice(0, op.wires[0], n_wires)
    181 sl_1 = _get_slice(1, op.wires[0], n_wires)
--> 183 state1 = math.multiply(state[sl_1], -1)
    184 return math.stack([state[sl_0], state1], axis=op.wires[0])

File /lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py:153, in filter_traceback.<locals>.error_handler(*args, **kwargs)
    151 except Exception as e:
    152   filtered_tb = _process_traceback_frames(e.__traceback__)
--> 153   raise e.with_traceback(filtered_tb) from None
    154 finally:
    155   del filtered_tb

File lib/python3.10/site-packages/tensorflow/python/framework/ops.py:7215, in raise_from_not_ok_status(e, name)
   7213 def raise_from_not_ok_status(e, name):
   7214   e.message += (" name: " + name if name is not None else "")
-> 7215   raise core._status_to_exception(e) from None

UnimplementedError: {{function_node __wrapped__StridedSlice_device_/job:localhost/replica:0/task:0/device:CPU:0}} Unhandled input dimensions 10 [Op:StridedSlice] name: strided_slice
```

 Since this case is exceptional, and we do not want to slow down the standard logical slow, I implemented fallback logic using try-except blocks.

If the error is a `tf.errors.UnimplementedError`, we fall back to `apply_operation_tensordot`.

This will probably have performance implications for applying PauliZ and CNOT on large states, but at least it won't affect the performance in cases where the error does not occur.

Blocks testing for #3862

### CODE REVIEW

**Comment by timmysilv** [comment]
imo, this can just be a conditional check right at the start of the function:
```python
n_dim = qml.math.ndim(state)
if n_dim >= 9 and qml.math.get_interface(state) == "tensorflow":
    return apply_operation_tensordot(op, state)
```
I think this is better for a few reasons:
- For smaller states, this will evaluate to false immediately without dispatching to `math`. For larger states, this won't be the bottleneck anyway
- The code is much clearer
- This only checks for the exact scenario that you know causes the bug. If something we didn't foresee happens, we might silently default to tensordot without realizing or knowing why
- I suspected that try-except is slow/not preferred in general, so I benchmarked the difference between try-except and using get_interface (with the lazy evaluation suggested above) over a range of system sizes. I attached a text file with results, but tl;dr the time is almost identical for low systems (try-except might be the teeny tiniest bit faster below 9 qubits because of the get_interface call, but it's a matter of microseconds), but notably faster for 9+ qubit systems, around 100 microseconds faster.

**Comment by albi3ro** [comment]
@timmysilv Thanks for running those benchmarks. I've updated the code to to check first.

File: pennylane/devices/qubit/apply_operation.py [hunk 1]
```diff
@@ -218,6 +224,10 @@ def apply_cnot(op: qml.CNOT, state, is_state_batched: bool = False):
     control_axes = op.wires[0] + is_state_batched
     n_dim = math.ndim(state)
 
+    n_dim = math.ndim(state)
```
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:227]:
```suggestion
```
Same suggestion as before

**Review by mudit2812 — COMMENTED** [review]
Looks good Christina! Are there any other interfaces such as torch that might also share the slicing trouble?

**Comment by albi3ro** [comment]
[sc-34873]

**Comment by albi3ro** [comment]
> Looks good Christina! Are there any other interfaces such as torch that might also share the slicing trouble?

This problem was found when writing tests for #3862 .  All other interfaces passed those tests.

**Review by mudit2812 — APPROVED** [review]
Thanks Christina!

**Review by timmysilv — APPROVED** [review]

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -111,7 +111,10 @@
 
 * Made `qml.OrbitalRotation` and consequently `qml.GateFabric` consistent with the interleaved Jordan-Wigner ordering.
   [(#3861)](https://github.com/PennyLaneAI/pennylane/pull/3861)
-
+  
+* `qml.devices.qubit.apply_operation` catches the `tf.errors.UnimplementedError` that occurs when `PauliZ` or `CNOT` gates
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:115]: i realize this isn't the most fitting changelog entry anymore... maybe let's sneak a touch-up into another PR?

**Pull request merged**